openapi/nexus.json

```diff
@@ -13717,13 +13717,17 @@
             "type": "string",
             "format": "uuid"
           },
+          "online_logical_cpus": {
```

Comment:
I've been struggling with the nomenclature here. This is probably fine, but I'd like to write up the different thoughts here as part of trying to think about future things. Part of my concern is what does the term cpu or processor mean and trying to come up with something that is understandable to lots of people, but I suspect this is a fool's errand in the end.

Fundamentally in the world that we are operating in, we have CPUs (referring in this instance to thing that we buy from someone) that are installed into sockets. They have some number of cores and some number of logical units beneath that that represent schedulable work (trying to avoid terms for a moment). Those logical units are called different things. For example Intel calls it "hyper-threading" and sometimes threads. In cpuid leaf 0xb and 0x1f they describe it as SMT. The AMD PPR is more willing to use the term threads in family 0x17 and 0x19. The amd64 level calls them threads.

But amd64 also had a thing they did in family 0x15 that lead the next level to be "logical cores" because of how they used CMT (clustered multi-threading). Which means the question of what is or isn't a core is weird, though this also isn't a CPU or thing that we expect to see again.

What we currently call a logical CPU corresponds to threads, which is what most folks seem to agree on. I've tried to look this up in ARM, but don't see great consensus. RISC-V does call this same thing a HART, or hardware thread, an says a core is made up of one or more HARTs.

From a provisioning perspective in the short and longer term we're interested in how many usable logical CPUs and cores there are. From an operator inventory perspective and overall capacity planning (versus in the moment), where we assume all hardware is usable, the things I think about are: asking how many sockets does a server have (more informational), how many total cores are there and how many threads/logical CPUs here are.

Finally, the last thing with nomenclature here is what do we think of when we're provisioning guests, which is the fundamental customer-facing abstraction. Right now the API file describes folks as passing `ncpus` and defines it as a CPU which right now maps to logical CPUs / threads in the system.

So playing this out, it seems like we'd expect a server object if we continue this trend to have:

* online_logical_cpus (still not sure if we should maybe call this cores)
* online_physical_cores
* total_logical_cpus
* total_physical_cpus
* total_physical_sockets

We may opt to leave out sockets here because it's not immediately useful. It's also worth noting that in a core-aware scheduling world, `online_logical_cpus` is not the same as what's usable. It's also not clear to me what we want to present with respect to what the actual provisionable CPU is from an operator's perspective, that is what is usable for guests, which is something that they care about, because it's probably not going to be all 128 even on a good day. This almost suggests hat we want a third category of `usable_xxx`, though this is starting to get complicated and more nuanced. Anyways, that's probably more on this than is useful.

Reply:
I'm worried that if we get too nuanced here it'll make the API harder to document and to reason about. I don't want to derail progress and perhaps it's something we can just put a pin in, but do other platforms define these terms in some way we could borrow? What would an operator typically see if they were using VMWare or Open stack or something? 

I have an uninformed perspective of this so my opinion counts for little here, but I wouldn't understand what `online_logical_cpus` means without reading documentation. Does `online` imply that there are `offline` versions too?

Reply:
Thanks for the detailed info!